### PR TITLE
Treat the absence of a truthy status code as an error

### PR DESCRIPTION
**Pull request by gwicke**
This was inspired by https://phabricator.wikimedia.org/T121611.

### CODE REVIEW

**Comment by Pchelolo** [comment]
@gwicke I've made a couple of experiments related to this. If we set a status code to some not-integer value, we'll get a 504 error from node http parser, for example for `unknown` or `5unknown` strings. However, if the number and a string is separated by a space, like in our case (`5 unknown`) the node http parser will parse it to integer 5. So this patch will not help in this particular case. May be we should assert that status is > 100? 

Anyway it should be fixed in parsoid too.

**Comment by gwicke** [comment]
@Pchelolo: That's.. subtle. Nice work!

File: index.js [hunk 1]
```diff
@@ -179,7 +179,7 @@ Request.prototype.run = function () {
                 }
             }
 
-            if (res.status >= 400) {
+            if (!res.status || res.status < 100 || res.status >= 400) {
```
Comment (d00rman) [inline index.js]: Shouldn't the code also set an appropriate status code in case it's missing or is `< 100` ? In this block we're just re-throwing it.

Comment (gwicke) [inline index.js]: Which status code would that be? 

I think from a user's perspective not inventing one is okay. We normally handle only specific errors, and bubbling random breakage like this up is the right thing in the common case.

Comment (Pchelolo) [inline index.js]: @gwicke IMHO a general 500 with message like `Response status parse error` could be fine, or we could mimic node's behaviour and throw 504, however the choice is not clear to me, as this has nothing to do with a `Gateway Timeout`

Comment (gwicke) [inline index.js]: The 504 is already something [we invented](https://github.com/gwicke/preq/blob/master/index.js#L198), to mark request timeouts with a status code. It's a bit of a stretch, but I think is still defendable.

For random invalid responses something in the 5xx range seems to make sense, but 500 would mix this with actual 500s returned by the service, which would be confusing. I don't see any official 5xx response code for invalid responses, so we'd probably have to invent one.

**Comment by gwicke** [comment]
@Pchelolo @eevans: ping

**Comment by Pchelolo** [comment]
@gwicke I still think, that we need to set some valid code in that case and not pass-through the invalid one.

**Comment by eevans** [comment]
I think I'm inclined to agree that the best course of action would have been to bubble up any random breakage; This being an HTTP client abstraction, it seems a little presumptuous to be supplying an HTTP status other than what the server returned (and it may violate the element of least surprise).

**Comment by gwicke** [comment]
Preq should provide sane & predictable error handling. The main mechanism it uses to report regular HTTP errors is to throw `HTTPError` instances, which contain details about the specific failure. This lets clients handle all HTTP errors systematically, and distinguish them from random preq exceptions / bugs.

So, I still think we should make HTTP error detection more robust, as is done in this patch. Whether we should set a 5xx status (like 502) in those responses is certainly debatable. There are pros & cons to either approach. Either way, I think the key part is more about systematically detecting & reporting invalid responses. Whether to set a status code & which one to set is somewhat secondary to that.

So, ping @wikimedia/services.

**Comment by d00rman** [comment]
Wow, this PR slipped through the cracks :-1: 

I agree wrt the need for more robust error detection. @eevans even though you are right that this is a client, I think that a server not setting a status is definitely a server error. Setting the code in preq allows clients to check the outcome more easily and alleviates the need for them to check this special case.

Now, whether 502 is the right way or not is a different question, but I think we can tweak it along the way. Perhaps we should invent another code? Something that says _Malformed response received from server_ or something similar?

**Pull request closed without merging**